File: tests/blink.ts
import { ByteStream, Byte, periodic } from '@amnisio/rivulet';
import { Sources, Sinks, HIGH, LOW, run, createSinks } from '@amnisio/arduino-uno';

const toggle = (value: Byte) => value == HIGH ? LOW : HIGH;

const blink = (arduino: Sources) => {
  const sinks = createSinks();
  sinks.LED$ =
    periodic(1000)
      .sample(arduino.LED$)
      .map(toggle);
  return sinks;
}

run(blink);

File: tests/blink.c


#include <Arduino.h>
#include <Gyrus.h>

Byte invert(Byte value) {
  return value == LOW ? HIGH : LOW;
}

Sinks* application(Sources* arduino) {
  Sinks* sinks = sinks_create();
  ByteStream* _typewriter_intermediary_1 = byte_stream_periodic(100);
  ByteStream* _typewriter_intermediary_2 = _typewriter_intermediary_1->sample(_typewriter_intermediary_1, arduino->D2);
  sinks->D2 = _typewriter_intermediary_2->map(_typewriter_intermediary_2, invert);
  ByteStream* _typewriter_intermediary_3 = byte_stream_periodic(250);
  ByteStream* _typewriter_intermediary_4 = _typewriter_intermediary_3->sample(_typewriter_intermediary_3, arduino->D3);
  sinks->D3 = _typewriter_intermediary_4->map(_typewriter_intermediary_4, invert);
  sinks->LED = byte_stream_periodic(500)->sample(byte_stream_periodic(500), arduino->LED)->map(byte_stream_periodic(500)->sample(byte_stream_periodic(500), arduino->LED), invert);
  return sinks;
}

void setup() {
  gyrus_run(application);
}

void loop() {
	
}
